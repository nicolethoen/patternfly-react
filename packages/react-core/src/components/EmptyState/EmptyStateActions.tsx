import * as React from 'react';
import { css } from '@patternfly/react-styles';
import styles from "@patternfly/react-styles/css/components/EmptyState/empty-state";

export interface EmptyStateActionsProps extends React.HTMLProps<HTMLDivElement> {
  /** Content rendered inside the EmptyStateActions */
  children?: React.ReactNode;
  /** Additional classes added to the EmptyStateActions */
  className?: string;
}

export const EmptyStateActions: React.FunctionComponent<EmptyStateActionsProps> = ({
  children = null,
  className = '',
  ...props
}: EmptyStateActionsProps) => (
  <div className={css(styles.emptyStateActions, className)} {...props}>
    {children}
  </div>
);
EmptyStateActions.displayName = 'EmptyStateActions';
