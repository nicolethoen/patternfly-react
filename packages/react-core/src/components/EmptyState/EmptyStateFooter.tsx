import * as React from 'react';
import { css } from '@patternfly/react-styles';
import styles from '@patternfly/react-styles/css/components/EmptyState/empty-state';

export interface EmptyStateFooterProps extends React.HTMLProps<HTMLDivElement> {
  /** Content rendered inside the EmptyStateFooter */
  children?: React.ReactNode;
  /** Additional classes added to the EmptyStateFooter */
  className?: string;
}

export const EmptyStateFooter: React.FunctionComponent<EmptyStateFooterProps> = ({
  children,
  className = '',
  ...props
}: EmptyStateFooterProps) => (
  <div className={css(styles.emptyStateFooter, className)} {...props}>
    {children}
  </div>
);
EmptyStateFooter.displayName = 'EmptyStateFooter';
