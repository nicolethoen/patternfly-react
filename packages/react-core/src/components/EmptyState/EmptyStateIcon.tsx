import * as React from 'react';
import { css } from '@patternfly/react-styles';
import styles from '@patternfly/react-styles/css/components/EmptyState/empty-state';
import { Spinner } from '../Spinner';

export interface IconProps extends Omit<React.HTMLProps<SVGElement>, 'size'> {
  /** Changes the color of the icon.  */
  color?: string;
}

export interface EmptyStateIconProps extends IconProps {
  /** Additional classes added to the empty state icon */
  className?: string;
  /** Icon component to be rendered inside the empty state on icon variant
   * Usually a CheckCircleIcon, ExclamationCircleIcon, LockIcon, PlusCircleIcon, RocketIcon
   * SearchIcon, or WrenchIcon. Can be a spinner component too.*/
  icon: React.ComponentType<any>;
}

const isSpinner = (icon: React.ReactElement<any>) => icon.type === Spinner;

export const EmptyStateIcon: React.FunctionComponent<EmptyStateIconProps> = ({
  className,
  icon: IconComponent,
  ...props
}: EmptyStateIconProps) => {
  return isSpinner(<IconComponent />) ? (
    <IconComponent className={css(className)}/>
  ) : (
    <div className={css(styles.emptyStateIcon, className)}>
      <IconComponent {...props} aria-hidden="true" />
    </div>
  );
}
EmptyStateIcon.displayName = 'EmptyStateIcon';
