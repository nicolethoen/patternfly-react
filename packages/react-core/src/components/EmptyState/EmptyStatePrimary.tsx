import * as React from 'react';
import { css } from '@patternfly/react-styles';

export interface EmptyStatePrimaryProps extends React.HTMLProps<HTMLDivElement> {
  /** Additional classes added to the empty state primary */
  className?: string;
  /** Content rendered inside the empty state primary */
  children: React.ReactNode;
}

export const EmptyStatePrimary: React.FunctionComponent<EmptyStatePrimaryProps> = ({
  children,
  className,
  ...props
}: EmptyStatePrimaryProps) => (
  <div className={css(className)} {...props}>
    {children}
  </div>
);
EmptyStatePrimary.displayName = 'EmptyStatePrimary';
