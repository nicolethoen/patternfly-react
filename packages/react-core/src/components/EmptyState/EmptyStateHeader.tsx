import * as React from 'react';
import { css } from '@patternfly/react-styles';
import styles from '@patternfly/react-styles/css/components/EmptyState/empty-state';
import { EmptyStateIcon } from './EmptyStateIcon';

export interface EmptyStateHeaderProps extends React.HTMLProps<HTMLDivElement> {
  /** Content rendered inside the empty state header */
  children?: React.ReactNode;
  /** Additional classes added to the empty state header */
  className?: string;
  /** Additional classes added to the title inside empty state header */
  titleClassName?: string;
  /** Additional classes added to the icon inside empty state header */
  iconClassName?: string;
  /** Text of the title inside empty state header */
  titleText?: string;
  /** Icon component to be rendered. Can also be a spinner component.  */
  icon?: React.ComponentType<any>;
  /** The heading level to use */
  headingLevel: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';
}

export const EmptyStateHeader: React.FunctionComponent<EmptyStateHeaderProps> = ({
  children,
  className,
  titleClassName,
  iconClassName,
  titleText,
  headingLevel: HeadingLevel = 'h1',
  icon,
  ...props
}: EmptyStateHeaderProps) => (
  <div className={css('pf-c-empty-state__header', className)} {...props}>
    {icon && <EmptyStateIcon className={iconClassName} icon={icon} />}
    {titleText && (
      <div className={css('pf-c-empty-state__title')}>
        <HeadingLevel className={css(styles.emptyStateTitleText, titleClassName)}>{titleText}</HeadingLevel>
      </div>
    )}
    {children}
  </div>
);
EmptyStateHeader.displayName = 'EmptyStateHeader';
