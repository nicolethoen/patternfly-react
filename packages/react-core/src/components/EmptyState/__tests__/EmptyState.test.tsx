import React from 'react';

import { render, screen } from '@testing-library/react';

import AddressBookIcon from '@patternfly/react-icons/dist/esm/icons/address-book-icon';
import { EmptyState, EmptyStateVariant } from '../EmptyState';
import { EmptyStateBody } from '../EmptyStateBody';
import { EmptyStateActions } from '../EmptyStateActions';
import { EmptyStateIcon } from '../EmptyStateIcon';
import { Button } from '../../Button';
import { EmptyStateHeader } from '../EmptyStateHeader';
import { EmptyStateFooter } from '../EmptyStateFooter';

describe('EmptyState', () => {
  test('Main', () => {
    const { asFragment } = render(
      <EmptyState>
        <EmptyStateHeader titleText="HTTP Proxies" />
        <EmptyStateBody>
          Defining HTTP Proxies that exist on your network allows you to perform various actions through those proxies.
        </EmptyStateBody>
        <EmptyStateFooter>
          <EmptyStateActions>
            <Button variant="primary">New HTTP Proxy</Button>
          </EmptyStateActions>
          <EmptyStateActions>
            <Button variant="link" aria-label="learn more action">
              Learn more about this in the documentation.
            </Button>
          </EmptyStateActions>
        </EmptyStateFooter>
      </EmptyState>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('Main variant large', () => {
    const { asFragment } = render(
      <EmptyState variant={EmptyStateVariant.large}>
        <EmptyStateHeader titleText="EmptyState large" />
      </EmptyState>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('Main variant small', () => {
    const { asFragment } = render(
      <EmptyState variant={EmptyStateVariant.small}>
        <EmptyStateHeader titleText="EmptyState small" />
      </EmptyState>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('Main variant xs', () => {
    const { asFragment } = render(
      <EmptyState variant={EmptyStateVariant.xs}>
        <EmptyStateHeader titleText="EmptyState small" />
      </EmptyState>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('Body', () => {
    render(<EmptyStateBody className="custom-empty-state-body" data-testid="body-test-id" />);
    expect(screen.getByTestId('body-test-id')).toHaveClass('custom-empty-state-body pf-c-empty-state__body');
  });

  xtest('Actions', () => {
    render(<EmptyStateActions className="custom-empty-state-secondary" data-testid="actions-test-id" />);
    expect(screen.getByTestId('actions-test-id')).toHaveClass('pf-c-empty-state__actions');
  });

  test('Icon', () => {
    render(<EmptyStateIcon icon={AddressBookIcon} data-testid="icon-test-id" />);
    expect(screen.getByTestId)
    expect(screen.getByTestId('icon-test-id').parentNode).toHaveClass('pf-c-empty-state__icon');
  });

  test('Wrap icon in a div', () => {
    const { container } = render(
      <EmptyStateIcon icon={AddressBookIcon} className="custom-empty-state-icon" id="empty-state-icon-id" />
    );

    expect(container.querySelector('div')).toHaveClass('pf-c-empty-state__icon custom-empty-state-icon');
    expect(container.querySelector('svg')).toBeInTheDocument();
  });

  test('Full height', () => {
    const { asFragment } = render(
      <EmptyState isFullHeight variant={EmptyStateVariant.large}>
        <EmptyStateHeader titleText="EmptyState large" />
      </EmptyState>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  // TODO rewrite tests for EmptyStateHeader -- icon and title to use React testing library
  // https://github.com/patternfly/patternfly-react/wiki/React-Testing-Library-Basics,-Best-Practices,-and-Guidelines
  test('Header', () => {
    const { container } = render(<EmptyStateHeader titleText='Empty state' icon={AddressBookIcon} />);

    expect(container.querySelector('h1')).toBeInTheDocument();
    expect(container.querySelector('h1')?.textContent).toBe('Empty state');
    expect(container.querySelector('svg')).toBeInTheDocument();
  });

  //TODO write tests for EmptyStateFooter
});
