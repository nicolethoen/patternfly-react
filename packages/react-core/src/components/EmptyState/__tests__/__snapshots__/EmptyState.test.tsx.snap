// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmptyState Full height 1`] = `
<DocumentFragment>
  <div
    class="pf-c-empty-state pf-m-lg pf-m-full-height"
  >
    <div
      class="pf-c-empty-state__content"
    >
      <div
        class="pf-c-empty-state__header"
      >
        <div
          class="pf-c-empty-state__title"
        >
          <h1
            class="pf-c-empty-state__title-text"
          >
            EmptyState large
          </h1>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`EmptyState Main 1`] = `
<DocumentFragment>
  <div
    class="pf-c-empty-state"
  >
    <div
      class="pf-c-empty-state__content"
    >
      <div
        class="pf-c-empty-state__header"
      >
        <div
          class="pf-c-empty-state__title"
        >
          <h1
            class="pf-c-empty-state__title-text"
          >
            HTTP Proxies
          </h1>
        </div>
      </div>
      <div
        class="pf-c-empty-state__body"
      >
        Defining HTTP Proxies that exist on your network allows you to perform various actions through those proxies.
      </div>
      <div
        class="pf-c-empty-state__footer"
      >
        <div
          class="pf-c-empty-state__actions"
        >
          <button
            aria-disabled="false"
            class="pf-c-button pf-m-primary"
            data-ouia-component-id="OUIA-Generated-Button-primary-1"
            data-ouia-component-type="PF4/Button"
            data-ouia-safe="true"
            type="button"
          >
            New HTTP Proxy
          </button>
        </div>
        <div
          class="pf-c-empty-state__actions"
        >
          <button
            aria-disabled="false"
            aria-label="learn more action"
            class="pf-c-button pf-m-link"
            data-ouia-component-id="OUIA-Generated-Button-link-1"
            data-ouia-component-type="PF4/Button"
            data-ouia-safe="true"
            type="button"
          >
            Learn more about this in the documentation.
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`EmptyState Main variant large 1`] = `
<DocumentFragment>
  <div
    class="pf-c-empty-state pf-m-lg"
  >
    <div
      class="pf-c-empty-state__content"
    >
      <div
        class="pf-c-empty-state__header"
      >
        <div
          class="pf-c-empty-state__title"
        >
          <h1
            class="pf-c-empty-state__title-text"
          >
            EmptyState large
          </h1>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`EmptyState Main variant small 1`] = `
<DocumentFragment>
  <div
    class="pf-c-empty-state pf-m-sm"
  >
    <div
      class="pf-c-empty-state__content"
    >
      <div
        class="pf-c-empty-state__header"
      >
        <div
          class="pf-c-empty-state__title"
        >
          <h1
            class="pf-c-empty-state__title-text"
          >
            EmptyState small
          </h1>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`EmptyState Main variant xs 1`] = `
<DocumentFragment>
  <div
    class="pf-c-empty-state pf-m-xs"
  >
    <div
      class="pf-c-empty-state__content"
    >
      <div
        class="pf-c-empty-state__header"
      >
        <div
          class="pf-c-empty-state__title"
        >
          <h1
            class="pf-c-empty-state__title-text"
          >
            EmptyState small
          </h1>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
